### PR TITLE
feat: support `farm` plugin

### LINKED ISSUE #340: [Feature Request]: farm plugin support
### Describe the feature

[github farm](https://github.com/farm-fe/farm)

[Farm Core Architecture Design js plugin](https://github.com/farm-fe/rfcs/blob/main/rfcs/001-core-architecture/rfc.md#62-js-plugin)

The farm plugin are inspired by rollup, So this might be straightforward to implement. I also tried to transplant to unplugin to implement all the hooks. With your permission, I'd like to do it.

### Additional information

- [X] Would you be willing to help implement this feature?

### PR DESCRIPTION
**Pull request by ErKeLost**
### 🔗 Linked issue
close #340 

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,25 +13,26 @@ Currently supports:
 - [Webpack](https://webpack.js.org/)
 - [esbuild](https://esbuild.github.io/)
 - [Rspack](https://www.rspack.dev/) (⚠️ experimental)
+- [Farm](https://farm-fe.github.io/)
```
Comment (antfu) [inline README.md:16]: Should we mark it as experimental?

File: README.md [hunk 2]
```diff
@@ -166,19 +174,37 @@ import { build } from 'esbuild'
 
 build({
   plugins: [
-    require('./unplugin-feature').esbuild({ /* options */ }),
+    require('./unplugin-feature').esbuild({
+      /* options */
+    }),
```
Comment (antfu) [inline README.md:179]: Could you revert the formatting changes here? Thanks.

Comment (ErKeLost) [reply README.md]: Formatting problem has been fixed

File: README.md (hunk 1 above)
Comment (ErKeLost) [reply README.md]: Farm has not yet released 1.0. it may follow the use of users to modify the plugin system. Thank you for your reminder.

**Comment by ErKeLost** [comment]
Inquiries about my Pull Request merger

hi . I'm sorry to bother you whether this pr can be merged, I think we are ready, unplugin will be an important part of the ecological environment of farm, or do you have any concerns ?

If you need any additional information or assistance to review my PR, I will try my best to provide it. I appreciate your time and efforts, as well as your support for farm.

Thank you!  @antfu

**Comment by antfu** [comment]
Hey, sorry for the late rely. I heard that Farm has become Vite-compactible to some extent. Is this PR still relevant? I am ok with having farm supported if there is anything more specific than Vite plugin, but hope the complexity included here could be slim down a bit as we won't have a lot of knowledge to maintain it in the long term. Thanks

**Comment by ErKeLost** [comment]
Thank you very much for your reply. Farm is currently compatible with vite Ecology. Our successful adaptation such as `unocss` means that we have achieved success in compatible vite Ecology, but we cannot guarantee compatibility with all api of vite, so we prefer to use farm plug-in system to write plug-ins. We will release 1.0 soon. Before that, I will modify the current pr to reduce the complexity and maintainability of the code as much as possible, and I am happy to join the unplugin family. Thanks

**Comment by antfu** [comment]
Great, looking forward to that! Converted this PR to draft, feel free to mark it as ready when you feel it's good to merge. Thanks

**Comment by ErKeLost** [comment]
Hi @antfu (Farm)[https://farm-fe.github.io/]has released 1.0.0-beta! At this exciting moment, I think we are ready. I feel like we are ready to move forward.

File: src/farm/utils.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import path from 'node:path'
+import * as querystring from 'node:querystring'
+import type { TransformResult } from 'src/types'
```
Comment (antfu) [inline src/farm/utils.ts:3]: This need to be relative path

Comment (ErKeLost) [reply src/farm/utils.ts]: done

**Comment by ErKeLost** [comment]
The document has also been modified to complete the home page. I added the features of `rolldown` and `farm` I don't know if you agree to do this.

**Comment by antfu** [comment]
Looks good to me, thank you! @ErKeLost Is that ok for us to ping you when we have issues regarding the farm integration, if any, later on?

**Review by antfu — APPROVED** [review]

**Pull request merged**